arch/rv32i/src/pmp.rs

```diff
@@ -39,306 +35,427 @@ register_bitfields![u8,
+    const CONST_ASSERT_CHECK: ();
+
+    /// The number of TOR regions currently available for userspace memory
+    /// protection. Within `[0; MAX_REGIONS]`.
+    ///
+    /// The PMP implementation may provide less regions than indicated through
+    /// `MAX_REGIONS`, for instance when entries are enforced (locked) in
+    /// machine mode. The number of available regions may change at runtime. The
+    /// implementation is free to map these regions to arbitrary PMP entries
+    /// (and change this mapping at runtime), provided that they are enforced
+    /// when the hart is in user-mode, and other memory regions are generally
+    /// inaccessible when in user-mode.
+    ///
+    /// When allocating regions for kernel-mode protection, and thus reducing
+    /// the number of regions available to userspace, re-configuring the PMP may
+    /// fail. This is allowed behavior. However, the PMP must not remove any
+    /// regions from the user-mode current configuration while it is active
+    /// ([`TORUserPMP::enable_user_pmp`] has been called, and it has not been
+    /// disabled through [`TORUserPMP::disable_user_pmp`]).
+    fn available_regions(&self) -> usize;
+
+    /// Configure the user-mode memory protection.
+    ///
+    /// This method configures the user-mode memory protection, to be enforced
+    /// on a call to [`TORUserPMP::enable_user_pmp`].
+    ///
+    /// PMP implementations where configured regions are only enforced in
+    /// user-mode may re-configure the PMP on this function invocation and
+    /// implement [`TORUserPMP::enable_user_pmp`] as a no-op. If configured
+    /// regions are enforced in machine-mode (for instance when using an ePMP
+    /// with the machine-mode whitelist policy), the new configuration rules
+    /// must not apply until [`TORUserPMP::enable_user_pmp`].
+    ///
+    /// The tuples as passed in the `regions` parameter are defined as follows:
+    ///
+    /// - first value ([`TORUserPMPCFG`]): the memory protection mode as
+    ///   enforced on the region. A `TORUserPMPCFG` can be created from the
+    ///   [`mpu::Permissions`] type. It is in a format compatible to the pmpcfgX
+    ///   register, guaranteed to not have the lock (`L`) bit set, and
+    ///   configured either as a TOR region (`A = 0b01`), or disabled (all bits
+    ///   set to `0`).
+    ///
+    /// - second value (`*const u8`): the region's start addres. As a PMP TOR
+    ///   region has a 4-byte address granularity, this address is rounded down
+    ///   to the next 4-byte boundary.
+    ///
+    /// - third value (`*const u8`): the region's end addres. As a PMP TOR
+    ///   region has a 4-byte address granularity, this address is rounded down
+    ///   to the next 4-byte boundary.
+    ///
+    /// To disable a region, set its configuration to [`TORUserPMPCFG::OFF`]. In
+    /// this case, the start and end addresses are ignored and can be set to
+    /// arbitrary values.
+    fn configure_pmp(
```

Comment:
Would is make sens to have these functions return specific error codes for the error conditions they might encounter?

Reply:
While I agree that we should ultimately have an error enumeration here, currently we can't even meaningfully handle these errors (the kernel's MPU trait and infrastructure assumes these operations to be infallible). I think that we should introduce more meaningful error variants once we change the general MPU interfaces to support erroring on an MPU configuration failure and acting accordingly (e.g., by suspending / failing the process instead of panicing).